sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark_source.py

```diff
@@ -0,0 +1,307 @@
+        """
+        return self._spark_options.path
+
+    @property
+    def file_format(self):
+        """
+        Returns the file format of this feature data source.
+        """
+        return self._spark_options.file_format
+
+    @staticmethod
+    def from_proto(data_source: DataSourceProto) -> Any:
+
+        assert data_source.HasField("custom_options")
+
+        spark_options = SparkOptions.from_proto(data_source.custom_options)
+        return SparkSource(
+            field_mapping=dict(data_source.field_mapping),
+            table=spark_options.table,
+            query=spark_options.query,
+            path=spark_options.path,
+            file_format=spark_options.file_format,
+            event_timestamp_column=data_source.event_timestamp_column,
+            created_timestamp_column=data_source.created_timestamp_column,
+            date_partition_column=data_source.date_partition_column,
+        )
+
+    def to_proto(self) -> DataSourceProto:
+        data_source_proto = DataSourceProto(
+            type=DataSourceProto.CUSTOM_SOURCE,
+            field_mapping=self.field_mapping,
+            custom_options=self.spark_options.to_proto(),
+        )
+
+        data_source_proto.event_timestamp_column = self.event_timestamp_column
+        data_source_proto.created_timestamp_column = self.created_timestamp_column
+        data_source_proto.date_partition_column = self.date_partition_column
+
+        return data_source_proto
+
+    def validate(self, config: RepoConfig):
+        self.get_table_column_names_and_types(config)
+
+    @staticmethod
+    def source_datatype_to_feast_value_type() -> Callable[[str], ValueType]:
+        return spark_to_feast_value_type
+
+    def get_table_column_names_and_types(
+        self, config: RepoConfig
+    ) -> Iterable[Tuple[str, str]]:
+        from feast.infra.offline_stores.third_party.spark_offline_store.spark import (
+            get_spark_session_or_start_new_with_repoconfig,
+        )
+
+        spark_session = get_spark_session_or_start_new_with_repoconfig(
+            store_config=config.offline_store
+        )
+        df = spark_session.sql(f"SELECT * FROM {self.get_table_query_string()}")
+        return (
+            (fields["name"], fields["type"])
+            for fields in df.schema.jsonValue()["fields"]
+        )
+
+    def get_table_query_string(self) -> str:
+        """Returns a string that can directly be used to reference this table in SQL"""
+        if self.table:
+            return f"`{self.table}`"
+        if self.query:
+            return f"({self.query})"
+
+        # If both the table query string and the actual query are null, we can load from file.
+        spark_session = SparkSession.getActiveSession()
+        if spark_session is None:
+            raise AssertionError("Could not find an active spark session.")
+        df = spark_session.read.format(self.file_format).load(self.path)
```

Comment:
should you catch any exception here (e.g. it isn't a file) so we can print a more helpful message in the exception?